pkg/flowaggregator/collector/ipfix.go

```diff
@@ -70,49 +90,119 @@ func NewIPFIXCollector(
 			IsEncrypted:   false,
 		}
 	default:
-		return nil, fmt.Errorf("unsupported protocol: %s", aggregatorTransportProtocol)
+		return fmt.Errorf("unsupported protocol: %s", c.aggregatorTransportProtocol)
 	}
 	// Tell the collector to accept IEs which are not part of the IPFIX registry (hardcoded in
 	// the go-ipfix library). The preprocessor will take care of removing these elements.
 	cpInput.DecodingMode = ipfixcollector.DecodingModeLenientKeepUnknown
 	collectingProcess, err := ipfixcollector.InitCollectingProcess(cpInput)
 	if err != nil {
-		return nil, fmt.Errorf("failed to initialize IPFIX collector: %w", err)
+		return fmt.Errorf("failed to initialize IPFIX collector: %w", err)
 	}
 
-	preprocessor, err := newPreprocessor(collectingProcess.GetMsgChan(), recordCh)
+	preprocessor, err := newPreprocessor(collectingProcess.GetMsgChan(), c.recordCh)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create IPFIX preprocessor: %w", err)
+		return fmt.Errorf("failed to create IPFIX preprocessor: %w", err)
 	}
 
-	return &ipfixCollector{
-		collectingProcess: collectingProcess,
-		preprocessor:      preprocessor,
-	}, nil
+	c.collectingProcess = collectingProcess
+	c.preprocessor = preprocessor
+
+	return nil
+}
+
+func (c *ipfixCollector) UpdateCerts() {
+	c.restartCh <- struct{}{}
 }
 
 func (c *ipfixCollector) Run(stopCh <-chan struct{}) {
+	err := c.initialize()
+	if err != nil {
+		klog.ErrorS(err, "unable to initialize ipfix collector")
+		return
+	}
+
+	go func() {
+		for {
+			select {
+			case <-stopCh:
+				return
+			case <-c.restartCh:
+				c.collectorReadyCond.L.Lock()
+				c.waitForRestart = true
+				oldCollectingProcess := c.collectingProcess
+				if err := c.initialize(); err != nil {
+					klog.ErrorS(err, "unable to re-initialize ipfix collector")
+				}
+				c.waitForRestart = false
+				c.collectorReadyCond.Broadcast()
```

Comment:
It's very rare to actually need a condition variable in Go, and I think that's exemplified that we only have one instance of it in the entire Antrea codebase
It's hard to understand the purpose of the code without comments, but I don't get why this is so complex, as opposed to a simple `c.collectingProcess.Stop()`, `c.initialize()`, `c.collectingProcess.Start()` sequence?

Reply:
Yes, the mutex is required
I am not aware of any issue with the `Stop` implementation in go-ipfix, but you will probably be able to use the `/debug/pprof` endpoint to collect a goroutine profile if you need to find out which goroutines are blocked.

Reply:
pprof was a good direction, thanks! I was able to resolve my issue and simplified the update.